substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/RemoveRedundantClassInitClosure.java

```diff
@@ -0,0 +1,135 @@
+* Base purpose of this class is to ensure that for every class, which is initialized at runtime, DynamicHub.ensureInitialized
+* method is generated only once.
+* It's called from RemoveRedundantClassInitPhase.
+* */
+public class RemoveRedundantClassInitClosure extends ReentrantBlockIterator.BlockIteratorClosure<EconomicSet<JavaConstant>> {
+
+    private static final int WHICH_ARGUMENT_INDEX_IS_CLASS = 0;
+
+    private EconomicMap<InvokeWithExceptionNode, AbstractBeginNode> nodesToBeRemoved;
+    private EconomicMap<InvokeWithExceptionNode, AbstractBeginNode> nodesToRemain;
+
+    private final HostedMethod classInitMethod;
+
+    public RemoveRedundantClassInitClosure(HostedMethod classInitMethod) {
+        this.classInitMethod = classInitMethod;
+    }
+
+    public void removeRedundantNodes(RemoveRedundantClassInitClosure removeRedundantClassInitClosure, Block start) {
+        ReentrantBlockIterator.apply(removeRedundantClassInitClosure, start);
+    }
+
+    public EconomicMap<InvokeWithExceptionNode, AbstractBeginNode> getNodesToBeRemoved() {
+        return nodesToBeRemoved;
+    }
+
+    public EconomicMap<InvokeWithExceptionNode, AbstractBeginNode> getNodesToRemain() {
+        return nodesToRemain;
+    }
+
+    @Override
+    protected EconomicSet<JavaConstant> getInitialState() {
+        this.nodesToBeRemoved = EconomicMap.create(Equivalence.IDENTITY);
+        this.nodesToRemain = EconomicMap.create(Equivalence.IDENTITY);
+
+        return EconomicSet.create(Equivalence.IDENTITY);
+    }
+
+    // processBlock method detects redundant initialization nodes base od their java constant values,
+    // and store them into nodesToBeRemoved EconomicMap. Opposite of nodesToBeRemoved is nodesToRemain EconomicMap.
+    @Override
+    protected EconomicSet<JavaConstant> processBlock(Block block, EconomicSet<JavaConstant> currentState) {
+        for (Node node : block.getNodes()) {
+            if (node instanceof InvokeWithExceptionNode) {
+                InvokeWithExceptionNode invokeWithExceptionNode = (InvokeWithExceptionNode) node;
+                if (invokeWithExceptionNode.callTarget() != null && invokeWithExceptionNode.callTarget().targetMethod() != null &&
+                        classInitMethod == invokeWithExceptionNode.callTarget().targetMethod()) {
+
+                    ValueNode valueNode = invokeWithExceptionNode.callTarget().arguments().get(WHICH_ARGUMENT_INDEX_IS_CLASS);
+                    if (valueNode.isJavaConstant()) {
+                        JavaConstant javaConstant = valueNode.asJavaConstant();
+                        if (!currentState.contains(javaConstant)) {
+                            currentState.add(javaConstant);
+                            nodesToRemain.put(invokeWithExceptionNode, invokeWithExceptionNode.next());
+                        } else {
+                            nodesToRemain.removeKey(invokeWithExceptionNode);
```

Comment:
How can this node ever be in `nodesToRemain` already?

Reply:
My mistake. I removed it.